### PR TITLE
filter manager: ability to convert upstream Starttls to secure mode

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
filter manager: ability to convert upstream Starttls to secure mode

Additional Description:
This PR provides a signalling path from a downstream read filter to a terminal filter in order to convert StartTLS upstream transport socket from non-secure mode to secure mode. StartTLS transport socket is basically a wrapper around raw (cleartext socket) and TLS socket. StartTLS transport socket always starts in cleartext and in order to convert it to secure mode, `startSecureTransport` method must be called. The main use of StartTLS transport socket is implementation of Opportunistic TLS protocol, where the first few messages are exchanged in cleartext and then TLS handshake is done on the same TCP session. The first few cleartext messages are protocol specific. Postgres messages are different than for example SMTP messages. Therefore a Read downstream filter is needed to parse those protocol specific messages and recognize the need to start TLS. The problem is that the Read filter (postgres, smtp, mysql) does not have an access to upstream connection, where the StartTLS transport socket is attached. The only filter which has a knowledge about both downstream and upstream connections it the terminal filter. Therefore the signal must be somehow conveyed from one of the downstream Read filters to the terminal filter. The result of the operation (whether upstream startTLS has been successfully converted to secure mode) should be returned to the Read filter.
That problem was discussed in one of the community meetings. @mattklein123 suggested to extend network FilterManager's callbacks and this PR uses that idea. The network filter manager iterates over all downstream filters and calls `startUpstreamSecureTransport` until one of them returns `true`. The only terminal filter which implements `startUpstreamSecureTransport` method is tcp_proxy. All other filters use `startUpstreamSecureTransport` defined in base class which is empty/no-op. @ggreenway raised a concern that tcp_proxy may be used for tunnelling. To address that issue only non-http upstream has `startUpstreamSecureTransport` implemented. The idea is described below.

FM

I used a modified postgres filter to exercise the new signalling path and it works properly.

Risk Level: Low. 
Testing: Added unit and integration testing.
Docs Changes: No.
Release Notes: No
Platform Specific Features: No.
Part of #19527

### CODE REVIEW

**Comment by ggreenway** [comment]
@lizan please review, or unassign yourself and I'll find another assignee.

File: source/common/network/filter_manager_impl.cc
Comment (lizan) [inline source/common/network/filter_manager_impl.cc:93]: nit: compare with nullptr and this if can be combined with the one below:

`if (filter->filter_ != nullptr && filter->filter_->startUpstreamSecureTransport())`

File: source/server/api_listener_impl.h
Comment (lizan) [inline source/server/api_listener_impl.h:71]:
```suggestion
      IS_ENVOY_BUG("Unexpected call to startUpstreamSecureTransport");
```

**Review by lizan — COMMENTED** [review]
this is in the right shape, a couple nits

**Comment by lizan** [comment]
/assign-from @envoyproxy/envoy-maintainers

for non-Tetrands review

**Comment by phlax** [comment]
im probably not the best maintainer to be assigned 8/ - @ggreenway is it ok to assign you ?

File: source/common/network/filter_manager_impl.cc
Comment (cpakulski) [reply source/common/network/filter_manager_impl.cc]: Done. Thanks.

**Comment by RyanTheOptimist** [comment]
@ggreenway would you mind giving this a look now that @lizan is happy with it?

File: test/common/tcp_proxy/tcp_proxy_test.cc
Comment (ggreenway) [inline test/common/tcp_proxy/tcp_proxy_test.cc:1333]:
```suggestion
TEST_F(TcpProxyTest, UpstreamStartSecureTransport) {
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
LGTM modulo one typo

/wait

**Review by ggreenway — APPROVED** [review]

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (soulxu) [inline source/common/tcp_proxy/tcp_proxy.cc:602]: Emm... seems like there should be an assertion to ensure upstream_ isn't nullptr.

**Pull request merged**
